src/parser.c

```
@@ -1413,23 +1435,25 @@ static bool parse_type_params(rbs_parser_t *parser, rbs_range_t *rg, bool module
       param_range.end = parser->current_token.range.end;
 
       rbs_location_t *loc = rbs_location_new(ALLOCATOR(), param_range);
-      rbs_loc_alloc_children(ALLOCATOR(), loc, 5);
+      rbs_loc_alloc_children(ALLOCATOR(), loc, 6);
       rbs_loc_add_required_child(loc, INTERN("name"), name_range);
       rbs_loc_add_optional_child(loc, INTERN("variance"), variance_range);
       rbs_loc_add_optional_child(loc, INTERN("unchecked"), unchecked_range);
       rbs_loc_add_optional_child(loc, INTERN("upper_bound"), upper_bound_range);
+      rbs_loc_add_optional_child(loc, INTERN("lower_bound"), lower_bound_range);
       rbs_loc_add_optional_child(loc, INTERN("default"), default_type_range);
 
-      rbs_ast_type_param_t *param = rbs_ast_type_param_new(ALLOCATOR(), loc, name, variance, upper_bound, default_type, unchecked);
+      rbs_ast_type_param_t *param = rbs_ast_type_param_new(ALLOCATOR(), loc, name, variance, upper_bound, lower_bound, default_type, unchecked);
 
       rbs_node_list_append(*params, (rbs_node_t *) param);
 
       if (parser->next_token.type == pCOMMA) {
         rbs_parser_advance(parser);
-      }
-
-      if (parser->next_token.type == pRBRACKET) {
+      } else if (parser->next_token.type == pRBRACKET) {
         break;
+      } else {
+        rbs_parser_set_error(parser, parser->next_token, true, "expected ',' or ']' after type parameter, got %s", rbs_token_type_str(parser->next_token.type));
+        return false;
```

This change ensures the parser correctly validates syntax after type parameter bounds. Previously, expressions like `[T < String Integer]` would be silently accepted despite being invalid.